public/components/hooks/use_polling.ts

```diff
@@ -0,0 +1,122 @@
+  enableExponentialBackoff = false,
+  maxRetries = 10, // Add a limit to the number of retries to prevent infinite retries
+  maxInterval = 60000, // Add a maximum interval, e.g., 60 seconds, to ensure the interval doesn't become unmanageably large
+}: UsePollingProps<T>): UsePollingResponse<T> {
+  const [data, setData] = useState<T | null>(null); // The latest fetched data
+  const [interval, setIntervalTime] = useState(initialInterval); // Current polling interval
+  const [isPolling, setPolling] = useState(true); // State to control polling status
+  const [error, setError] = useState<any>(null); // Error state
+  const cache = useRef<Record<string, T>>({}); // Cache for storing intermediate data
+  const retries = useRef(0); // Number of retries after failed fetches
+  const isMounted = useRef(true); // To prevent state updates on unmounted components
+  const isPrevCallCompleted = useRef(true); // To prevent overlapping fetch calls
+
+  useEffect(() => {
+    // If polling is stopped, don't execute any fetch logic
+    if (!isPolling) return;
+
+    const doFetch = async () => {
+      // Prevent overlapping fetches
+      if (!isPrevCallCompleted.current) return;
+      isPrevCallCompleted.current = false;
+
+      try {
+        const result = await fetchFunction();
+
+        // Update state only if the component is still mounted
+        if (isMounted.current) {
+          setData(result);
+          setError(null); // Reset the error if the fetch is successful
+          retries.current = 0; // Reset retries on a successful fetch
+          setIntervalTime(initialInterval); // Reset interval on a successful fetch
+          // Update cache if a caching strategy is provided
+          if (cachingStrategy) {
+            cachingStrategy(result, cache.current);
+          }
+        }
+      } catch (err) {
+        setError(err);
+
+        // Handle exponential backoff if enabled
+        if (enableExponentialBackoff && retries.current < maxRetries) {
+          console.log('retries.current: ', retries.current);
+          retries.current += 1;
+          let nextInterval = initialInterval * 2;
+          nextInterval = Math.min(nextInterval, maxInterval); // Ensure we don't exceed the max interval
+          setIntervalTime(nextInterval);
+        }
+      } finally {
+        isPrevCallCompleted.current = true;
+      }
+    };
+
+    // Set up the polling interval
+    const intervalId = setInterval(doFetch, interval);
+
+    // Immediately call the fetch function once
+    doFetch();
+
+    // Cleanup: clear the interval and update the isMounted ref
+    return () => {
+      isMounted.current = false;
+      clearInterval(intervalId);
+    };
+  }, [
+    fetchFunction,
+    interval,
+    isPolling,
+    cachingStrategy,
+    initialInterval,
+    enableExponentialBackoff,
+  ]);
+
+  return {
+    data,
+    cache: cache.current,
+    startPolling: () => setPolling(true),
+    stopPolling: () => setPolling(false),
```

Comment:
stopPolling might work as it changes `isPolling` to trigger another useEffect, and the cleanup of the previous useEffect is called, clearing the interval. But `isMounted` is also set to false. when calling `startPolling` again, `isMounted` stays false and results are ignored?